Add dynamic_wk_cast to safely cast between different types of WKTypeRef

achristensen07 opened the pull request:
#### e15a031d20be12b21baf6663ae4e669cfc29bce0
<pre>
Add dynamic_wk_cast to safely cast between different types of WKTypeRef
<a href="https://bugs.webkit.org/show_bug.cgi?id=298285">https://bugs.webkit.org/show_bug.cgi?id=298285</a>

Reviewed by Simon Fraser.

* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/UIProcess/API/cpp/WKCast.h: Added.
(WebKit::wk_cast):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::findOptionsFromArray):
(WTR::CompletionHandler&lt;void):

Canonical link: <a href="https://commits.webkit.org/299503@main">https://commits.webkit.org/299503@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3e4835f66cde211c7ca5281e9dc53774eecf1355

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119246 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38929 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29583 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125477 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71311 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4032a00b-ee3a-4d14-ab27-6769518ce397) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121124 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39625 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47510 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/90596 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0b67ede7-09c8-48c6-ad63-59437ba44269) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122199 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31603 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106902 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71011 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/100a558d-cd94-444b-9825-ad4a1be538f3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30646 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25015 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69127 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101050 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25200 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128490 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46158 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34895 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/99161 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46523 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103112 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98939 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44406 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22412 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18976 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46024 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51708 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45490 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48840 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47180 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8269dbe7bd16c444b8b6e4f5f94bdf9da32bddbb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50216

cdumez commented on Source/WebKit/UIProcess/API/cpp/WKCast.h:
```diff
@@ -0,0 +1,70 @@
+
+template<typename T, typename U> inline T wk_cast(U* object)
+{
+    if (!object || WKGetTypeID(object) != detail::TypeCheckHelper<T>::typeID())
```
We already have downcast vs dynamicDowncast distinction. We also have `dynamic_objc_cast` for ObjC.
Here, since it early returns nullptr in case of type mismatch, I would expect this cast function to have "dynamic" in the name OR RELEASE_ASSERT that the type is correct.

achristensen07 replied:
dynamic_wk_cast

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/43aa551b2700dd75f84cd6a653d85445d762bf67. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50216

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4d97e2be814c261a2c760e991d51cf066bb3fd02. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50216

achristensen07 committed:
Add dynamic_wk_cast to safely cast between different types of WKTypeRef
https://bugs.webkit.org/show_bug.cgi?id=298285

Reviewed by Simon Fraser.

* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/UIProcess/API/cpp/WKCast.h: Added.
(WebKit::wk_cast):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::findOptionsFromArray):
(WTR::CompletionHandler<void):

Canonical link: https://commits.webkit.org/299503@main
